core/sail/inferencer/src/main/java/org/eclipse/rdf4j/sail/inferencer/fc/FastRdfsForwardChainingSail.java

```diff
@@ -0,0 +1,311 @@
+    Set<Resource> properties = new HashSet<>();
+    List<Statement> subPropertyOfStatements = new ArrayList<>();
+    List<Statement> rangeStatements = new ArrayList<>();
+    List<Statement> domainStatements = new ArrayList<>();
+
+
+    Map<Resource, Set<Resource>> calculatedTypes = new HashMap<>();
+    Map<Resource, Set<Resource>> calculatedProperties = new HashMap<>();
+    Map<Resource, Set<Resource>> calculatedRange = new HashMap<>();
+    Map<Resource, Set<Resource>> calculatedDomain = new HashMap<>();
+    private boolean sharedSchema;
+
+    void clearInferenceTables() {
+        subClassOfStatements = new ArrayList<>();
+        properties = new HashSet<>();
+        subPropertyOfStatements = new ArrayList<>();
+        rangeStatements = new ArrayList<>();
+        domainStatements = new ArrayList<>();
+        calculatedTypes = new HashMap<>();
+        calculatedProperties = new HashMap<>();
+        calculatedRange = new HashMap<>();
+        calculatedDomain = new HashMap<>();
+    }
+
+    void readLock(FastRdfsForwardChainingSailConnection connection) {
+
+        if (numberOfThreadsWaitingForWriteLock.get() > 0) {
+//            System.err.println("starve reads");
+        }
+
+        while (numberOfThreadsWaitingForWriteLock.get() > 0) {
+            Thread.yield();
+        }
+
+        if (connection.lockStamp != 0) {
+            throw new IllegalStateException("Connection already has a lock!");
+        }
+        connection.lockStamp = readWriteLock.readLock();
+//        System.err.println("readLock: " + connection.lockStamp);
+
+    }
+
+    void releaseLock(FastRdfsForwardChainingSailConnection connection) {
+        readWriteLock.unlock(connection.lockStamp);
+//        System.err.println("Released lock: " + connection.lockStamp);
+
+        connection.lockStamp = 0;
+    }
+
+
+    AtomicInteger numberOfThreadsWaitingForWriteLock = new AtomicInteger(0);
+
+    void upgradeLock(FastRdfsForwardChainingSailConnection connection) {
+
+////        System.err.println("Attempt writelock: "+connection.lockStamp);
+
+        numberOfThreadsWaitingForWriteLock.incrementAndGet();
+
+        try {
+           while(true) {
+                long l = readWriteLock.tryConvertToWriteLock(connection.lockStamp);
+
+                if (l != 0) {
+                    long temp = connection.lockStamp;
+                    connection.lockStamp = l;
+                    if (temp != l) {
+//                        System.err.println("readLock: " + temp + " writeLock: " + connection.lockStamp);
+                    }
+
+
+                    return;
+                }
+
+                try {
+                    Thread.sleep(random.nextInt(2));
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
+                }
+
+                // detect potential deadlock scenario
+                if(numberOfThreadsWaitingForWriteLock.get() > 1  //More than 1 connection waiting for a write lock
```

Comment:
That makes sense, but I'm uncomfortable with this notion of randomly selecting a transaction to abort. Why not  use `writelock().tryWriteLock(timeout)` with a sensible timeout and abort the _current_ transaction (rather than a random concurrent transaction) if that fails? The timeout would be a fixed estimate on how long any given operation will hold on to a read lock.

Scenario one: Thread B tries to obtain write lock - which blocks until read lock from A is released or timeout occurs. 

Scenario two: Thread A and B contest for the write lock, and end up in a temporary deadlock. Depending on which thread's try times out first, that thread fails, and the other can obtain the write lock.